# XGBoost Regression
This programm runs XGBoost regression to attempt to predict stock prices after three months.

### 1. Imports

In [1]:
import xgboost as xgb
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, r2_score

### 2. Load the data

In [2]:
data = pd.read_csv('stocks_data.csv')
data.describe(include='all')

Unnamed: 0 Ticker          Year         Month      MA Ratio  \
count   14854.000000  14854  14854.000000  14854.000000  14854.000000   
unique           NaN    393           NaN           NaN           NaN   
top              NaN    AWK           NaN           NaN           NaN   
freq             NaN     55           NaN           NaN           NaN   
mean     7426.500000    NaN   2020.669113      6.223240      1.004148   
std      4288.124784    NaN      1.428016      3.520757      0.046473   
min         0.000000    NaN   2018.000000      1.000000      0.580721   
25%      3713.250000    NaN   2019.000000      3.000000      0.977766   
50%      7426.500000    NaN   2021.000000      6.000000      1.005360   
75%     11139.750000    NaN   2022.000000      9.000000      1.031953   
max     14853.000000    NaN   2023.000000     12.000000      1.335656   

              Result           ROE  Insider Ownership Growth  \
count   14854.000000  14854.000000              14854.000000   
unique           NaN           NaN                       NaN   
top              NaN           NaN                       NaN   
freq             NaN           NaN                       NaN   
mean        1.032817     39.494365                  0.015486   
std         0.149260    181.839873                  0.269863   
min         0.259712   -613.743387                 -0.633527   
25%         0.944153     10.160854                 -0.001350   
50%         1.028547     19.251991                  0.000000   
75%         1.113949     31.949569                  0.008000   
max         3.142137  11300.330487                 17.759731   

        Institutional Ownership Growth  Forecast EPS Growth  \
count                     14854.000000         14854.000000   
unique                             NaN                  NaN   
top                                NaN                  NaN   
freq                               NaN                  NaN   
mean                          0.026708             0.057775   
std                           0.230675             2.136724   
min                          -0.714136            -0.992366   
25%                          -0.023114            -0.184264   
50%                          -0.000648            -0.039062   
75%                           0.033653             0.086957   
max                          14.203540           133.000000   

        Avg 2Q EPS Growth  Avg 2Q EPS Surprise  YoY EPS Growth  \
count        14854.000000         14854.000000    14854.000000   
unique                NaN                  NaN             NaN   
top                   NaN                  NaN             NaN   
freq                  NaN                  NaN             NaN   
mean             0.181477            13.755183        0.369529   
std              2.111809            46.751483        3.637998   
min            -45.050000           -65.625000       -0.961538   
25%             -0.040838             2.015000        0.017606   
50%              0.045662             6.055000        0.130688   
75%              0.154182            13.135000        0.275148   
max             78.553797          1493.215000      192.000000   

        Sector Performance  Market Performance  
count         14854.000000        14854.000000  
unique                 NaN                 NaN  
top                    NaN                 NaN  
freq                   NaN                 NaN  
mean              1.488003            1.438443  
std               8.164589            7.038394  
min             -44.900728          -22.795349  
25%              -3.453784           -3.160007  
50%               1.496227            2.069271  
75%               6.429508            5.507430  
max              31.971220           21.644623

### 3. Split and clean the data
Based on my experience from attempt using Random Forest Regression (see RandomForest.ipynb), where overfitting occurred, this time I will split the data according to time rather than randomly.

In [3]:
data = data.reset_index(drop=True)
train_data = data[data['Year'] < 2023]
test_data = data[data['Year'] >= 2023]
x_train = train_data.drop(['Year', 'Result', 'Month', 'Ticker', data.columns[0]], axis=1)
y_train = train_data['Result']
x_test = test_data.drop(['Year', 'Result', 'Month', 'Ticker', data.columns[0]], axis=1)
y_test = test_data['Result']

### 4. Run regression

In [4]:
xgb_regressor = xgb.XGBRegressor(objective='reg:squarederror', n_estimators=100, random_state=42)
xgb_regressor.fit(x_train, y_train)
y_pred = xgb_regressor.predict(x_test)

### 5. Evaluation

In [5]:
mse = mean_squared_error(y_test, y_pred)
print(f'Mean Squared Error: {mse}')
r2 = r2_score(y_test, y_pred)
print(f'R^2 Score: {r2}')

Mean Squared Error: 0.02983337233335683
R^2 Score: -0.33087834658024007


### 6. Conclusion
In this attempt with XGBoost regression, the results are as follows:
- Mean Squared Error (MSE): 0.0298
- R² Score: -0.3309

The negative R² score indicates that the model is performing significantly worse than a baseline prediction using the mean of the target values. This suggests that the model has struggled to capture the underlying patterns in the data. The poor performance is likely due to the limited dataset, which only covers a relatively short time span and includes just one recessionary period (2022). This lack of data diversity likely hindered the model's ability to learn and generalize, leading to overfitting or an inability to predict accurately for different time periods.

##### Next Steps
To improve the model's performance, it would be beneficial to expand the dataset to include a longer time frame with multiple economic cycles. This would provide more varied data, allowing the model to better capture the underlying trends and improve its generalization ability.